Extraction.unflatten() will throw MatchError when field name is prefix of the other field name.

### json4s version
3.6.7
### scala version
2.11.11
### jdk version
AdoptOpenJDK 1.8

I have noticed a probable bug  when trying to call `Extraction.unflatten()`. When we will call `unflatten` on previously flatenned JSON string in situtation where name of one field is prefix of the other fields name. So, let's say we have the following JSON :
```
{"data" : 5, "data_type": 6} 
```
If we will `flatten` this and then try to `unflatten`, it will throw `MatchError`. The same will happen if  instead of `data_type` we will have `data2`, `datad` or any kind of shated prefix. The exception for the descibed case is :
```
Caused by: scala.MatchError: _type (of class java.lang.String)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:351)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:350)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at org.json4s.Extraction$.unflatten(Extraction.scala:350)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:354)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:350)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at org.json4s.Extraction$.unflatten(Extraction.scala:350)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:354)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:350)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at org.json4s.Extraction$.unflatten(Extraction.scala:350)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:354)
	at org.json4s.Extraction$$anonfun$unflatten$1.apply(Extraction.scala:350)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:124)
	at scala.collection.immutable.List.foldLeft(List.scala:84)
	at org.json4s.Extraction$.unflatten(Extraction.scala:350)
```

--- a/core/src/main/scala/org/json4s/Extraction.scala
+++ b/core/src/main/scala/org/json4s/Extraction.scala
@@ -330,14 +330,20 @@ object Extraction {
         else JString(ParserUtil.unquote(value.substring(1)))
     }
 
-    def submap(prefix: String): Map[String, String] =
-      map.withFilter(t => t._1.startsWith(prefix)).map(
-        t => (t._1.substring(prefix.length), t._2)
-      )
-
     val ArrayProp = new Regex("""^(\.([^\.\[]+))\[(\d+)\].*$""")
     val ArrayElem = new Regex("""^(\[(\d+)\]).*$""")
     val OtherProp = new Regex("""^(\.([^\.\[]+)).*$""")
+
+    def submap(prefix: String): Map[String, String] = {
+      map.withFilter { t =>
+        t._1 match {
+          case ArrayProp(p, _, _) if p == prefix => true
+          case ArrayElem(p, _)    if p == prefix => true
+          case OtherProp(p, _)    if p == prefix => true
+          case _                  => false
+        }
+      } map { t => (t._1.substring(prefix.length), t._2) }
+    }
 
     val uniquePaths = map.keys.foldLeft[Set[String]](Set()) {
       (set, key) =>

--- a/tests/src/test/scala/org/json4s/ExtractionExamplesSpec.scala
+++ b/tests/src/test/scala/org/json4s/ExtractionExamplesSpec.scala
@@ -198,6 +198,12 @@ abstract class ExtractionExamples[T](mod: String, ser : json4s.Serialization) ex
       Extraction.unflatten(m) must_== JObject(List(JField("foo", JArray(List(JInt(0), JInt(1), JInt(2))))))
     }
 
+    "Unflatten example with field name is prefix of the other field name" in {
+      val m = Map(".data" -> "5", ".data_type" -> "6")
+
+      Extraction.unflatten(m) must_== JObject(JField("data", JInt(5)), JField("data_type", JInt(6)))
+    }
+
     "Flatten and unflatten are symmetric" in {
       val parsed = parse(testJson)
 

